File: leetcode/1_two_sum.rb
# Given an array of integers, return indices of the two numbers such that they add up to a specific target.
#
# You may assume that each input would have exactly one solution, and you may not use the same element twice.
#
# Example:
#
# Given nums = [2, 7, 11, 15], target = 9,
#
# Because nums[0] + nums[1] = 2 + 7 = 9,
# return [0, 1].

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  # instantiate dictionary
  dict = {}
  # loop through nums array with element and index
  nums.each_with_index do |n, i|
    # if the dictionary contains the complement then return the complement and current index
    if dict[target - n]
      return [dict[target - n], i]
    else
      # if there is no complement store it in our dictionary
      dict[n] = i
    end
  end
end

File: leetcode/1_two_sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

// Brute force -> O(N^2)
let twoSum = function (nums, target) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) {
        return [i, j];
      }
    }
  }
};

// O(N) hash map
var twoSum = function (nums, target) {
  let map = new Map();

  for (let i = 0; i < nums.length; i++) {
    let num = nums[i];
    // if there is no complement in the array then set it
    if (map.get(num) === undefined) {
      // set the complement and their indicy
      map.set(target - num, i);
    } else {
      // get both indicies of the complements that add up to the target
      return [map.get(num), i];
    }
  }
};
